### PR TITLE
Improve cache eviction policy for LoadedPrograms

### PR DESCRIPTION
**Pull request by pgarg66**
#### Problem
Current eviction policy for `LoadedPrograms` cache relies on program's `usage_counter` to decide which entry to evict. The `usage_counter` increases monotonically. This means if an old program which was frequently used in the past, but not used anymore, has higher chance of staying in the cache. This can make it harder for new programs to stay in the cache once the cache depth limits are reached. 

#### Summary of Changes
- Decay the `usage_counter` of a program if the program does not get used.
- Use 2's random selection to pick a program entry to be evicted from the cache.
- Add unit tests for the new code

Fixes #

### CODE REVIEW

**Comment by pgarg66** [comment]
Also, running a node against MB. No perceptible difference in eviction rate compared to the rest of the cluster.

File: program-runtime/src/loaded_programs.rs [hunk 1]
```diff
@@ -428,6 +435,19 @@ impl LoadedProgram {
             && slot >= self.deployment_slot
             && slot < self.effective_slot
     }
+
+    pub fn update_access_slot(&self, slot: Slot) {
+        let mut current_slot = self.latest_access_slot.write().unwrap();
```
Comment (t-nelson) [inline program-runtime/src/loaded_programs.rs:440]: could drop the lock in favor of another `AtomicU64`. this method is then a loop on `compare_exchange` while `new < current`

File: program-runtime/src/loaded_programs.rs [hunk 2]
```diff
@@ -914,6 +952,39 @@ impl<FG: ForkGraph> LoadedPrograms<FG> {
         self.unload_program_entries(sorted_candidates.iter().take(num_to_unload));
     }
 
+    /// Evicts programs using 2's random selection, choosing the least used program out of the two entries.
+    /// The eviction is performed enough number of times to reduce the cache usage to the given percentage.
+    pub fn evict_using_2s_random_selection(&mut self, shrink_to: PercentageInteger, now: Slot) {
+        let mut candidates = self.get_flattened_entries();
```
Comment (t-nelson) [inline program-runtime/src/loaded_programs.rs:958]: could switch `self.entries` from `HashMap` to `IndexMap`. then the draw is just two random samplings from `0..self.entries.len()`. no allocation necessary

File: program-runtime/src/loaded_programs.rs [hunk 3]
```diff
@@ -914,6 +952,39 @@ impl<FG: ForkGraph> LoadedPrograms<FG> {
...
+                })
+            {
+                let (program, entry) = if usage_counter1 < usage_counter2 {
+                    candidates.remove(index1)
+                } else {
+                    candidates.remove(index2)
+                };
+                self.unload_program_entry(&program, &entry);
+            } else {
+                error!("Failed in evicting an entry from the program cache");
```
Comment (t-nelson) [inline program-runtime/src/loaded_programs.rs:983]: in this case the cache is growing without bound?

when could the candidates be `None`?

File: program-runtime/src/loaded_programs.rs [hunk 4]
```diff
@@ -941,26 +1012,30 @@ impl<FG: ForkGraph> LoadedPrograms<FG> {
         keys.iter().for_each(|key| self.unload_program(key));
     }
 
+    fn unload_program_entry(&mut self, program: &Pubkey, remove_entry: &Arc<LoadedProgram>) {
+        if let Some(entries) = self.entries.get_mut(program) {
+            if let Some(candidate) = entries.iter_mut().find(|entry| entry == &remove_entry) {
+                if let Some(unloaded) = candidate.to_unloaded() {
```
Comment (t-nelson) [inline program-runtime/src/loaded_programs.rs:1018]: any of these being `None` means (and _meant_ prior to this change) that we're failing silently

File: program-runtime/src/loaded_programs.rs [hunk 5]
```diff
@@ -914,6 +952,39 @@ impl<FG: ForkGraph> LoadedPrograms<FG> {
         self.unload_program_entries(sorted_candidates.iter().take(num_to_unload));
     }
 
+    /// Evicts programs using 2's random selection, choosing the least used program out of the two entries.
+    /// The eviction is performed enough number of times to reduce the cache usage to the given percentage.
+    pub fn evict_using_2s_random_selection(&mut self, shrink_to: PercentageInteger, now: Slot) {
```
Comment (t-nelson) [inline program-runtime/src/loaded_programs.rs:957]: @behzadnouri mind giving this method a once over?

File: program-runtime/src/loaded_programs.rs [hunk 6]
```diff
@@ -1122,6 +1200,176 @@ mod tests {
...
+        assert_eq!(program.decayed_usage_counter(16), 32);
+        assert_eq!(program.decayed_usage_counter(17), 32);
+        assert_eq!(program.decayed_usage_counter(18), 32);
+        assert_eq!(program.decayed_usage_counter(19), 16);
+        assert_eq!(program.decayed_usage_counter(20), 8);
+        assert_eq!(program.decayed_usage_counter(21), 4);
+    }
+
+    #[test]
+    fn test_random_eviction() {
```
Comment (t-nelson) [inline program-runtime/src/loaded_programs.rs:1229]: this test is kinda hard to follow at a glance. is it testing the failure cases? i only see assertions for expected results, not that the setup proceeded accordingly or for individual properties under test

File: program-runtime/src/loaded_programs.rs (hunk 1 above)
Comment (pgarg66) [reply program-runtime/src/loaded_programs.rs:440]: I had first done it with `AtomicU64`, and then decided to use the absolute type for `Slot`. I don't think we ever will change `Slot` to not being a `u64`. So maybe it's ok to use `AtomicU64`.

File: program-runtime/src/loaded_programs.rs (hunk 2 above)
Comment (pgarg66) [reply program-runtime/src/loaded_programs.rs:958]: Changing to `IndexMap` will impact other parts of the code in the cache, not just eviction logic. If we agree to do it, maybe best to do it in a separate PR (in case we want to backport this). Thoughts?

File: program-runtime/src/loaded_programs.rs (hunk 3 above)
Comment (pgarg66) [reply program-runtime/src/loaded_programs.rs:983]: It should never be `None`, since the random index is within the range. I didn't want to do `unwrap()` or `expect()`. What would be a semantically better approach?

File: program-runtime/src/loaded_programs.rs (hunk 6 above)
Comment (pgarg66) [reply program-runtime/src/loaded_programs.rs:1229]: Agree that the test is a bit complicated. It's adding different kind of entries to the cache, some of which are expected to never be evicted. Also, it adds multiple entries for a given program. So it tries to create a typical cache instance. Most of the test code is actually constructing the cache.

The call to eviction would not fail in itself, i.e. it should evict the necessary entries. The test is ensuring that the number of entries in the cache after eviction matches the expected count.

File: program-runtime/src/loaded_programs.rs (hunk 4 above)
Comment (pgarg66) [reply program-runtime/src/loaded_programs.rs:1018]: I'll add some `error!` logs for these conditions.
Ideally it should never happen, since the entry was looked up before we tried to remove it.

Comment (t-nelson) [reply program-runtime/src/loaded_programs.rs:1034]: i imagine panicking is more appropriate given, as you said, if this happens, something is very wrong and it is unlikely that proceeding is safe

File: program-runtime/src/loaded_programs.rs (hunk 6 above)
Comment (t-nelson) [reply program-runtime/src/loaded_programs.rs:1257]: maybe some comments along those lines would help?

File: program-runtime/src/loaded_programs.rs (hunk 2 above)
Comment (pgarg66) [reply program-runtime/src/loaded_programs.rs:967]: Each entry in `entries` is a `Vec<>` of cached programs (since a given program ID can potentially have different versions of the program loaded simultaneously in cache corresponding to different slot/fork). So we need to flatten the two dimensional list to actually compute the total cache usage, and count how much to evict. I don't think we can avoid allocation even if we were to use IndexMap here.

File: program-runtime/src/loaded_programs.rs (hunk 5 above)
Comment (pgarg66) [reply program-runtime/src/loaded_programs.rs:966]: @Lichtso, @alessandrod?

File: program-runtime/src/loaded_programs.rs [hunk 7]
```diff
@@ -428,6 +435,28 @@ impl LoadedProgram {
             && slot >= self.deployment_slot
             && slot < self.effective_slot
     }
+
+    pub fn update_access_slot(&self, slot: Slot) {
+        let mut last_access = self.latest_access_slot.load(Ordering::Relaxed);
```
Comment (alessandrod) [inline program-runtime/src/loaded_programs.rs:440]: I think this is fetch_update?

File: program-runtime/src/loaded_programs.rs (hunk 2 above)
Comment (alessandrod) [reply program-runtime/src/loaded_programs.rs:967]: This sounds like a case for a BTreeMap? With keys being (program_key, slot) or something

File: program-runtime/src/loaded_programs.rs (hunk 5 above)
Comment (alessandrod) [reply program-runtime/src/loaded_programs.rs:966]: the index selection code lgtm

File: program-runtime/src/loaded_programs.rs (hunk 6 above)
Comment (alessandrod) [reply program-runtime/src/loaded_programs.rs:1251]: nit: usage counters from this test are always zero (probably because of decay?), so
it's not testing both branches of the random eviction (the code is correct, but
still :P)

```
cargo test -p solana-program-runtime -- test_random_eviction --nocapture
    Finished test [unoptimized + debuginfo] target(s) in 0.27s
     Running unittests src/lib.rs (target/debug/deps/solana_program_runtime-2b785cc50c25ca56)

running 1 test
[program-runtime/src/loaded_programs.rs:989] usage_counter1 = 0
[program-runtime/src/loaded_programs.rs:989] usage_counter2 = 0
[program-runtime/src/loaded_programs.rs:989] usage_counter1 = 0
[program-runtime/src/loaded_programs.rs:989] usage_counter2 = 0
[program-runtime/src/loaded_programs.rs:989] usage_counter1 = 0
[program-runtime/src/loaded_programs.rs:989] usage_counter2 = 0
test loaded_programs::tests::test_random_eviction ... ok
```

**Comment by alessandrod** [comment]
some comments

File: program-runtime/src/loaded_programs.rs (hunk 2 above)
Comment (alessandrod) [reply program-runtime/src/loaded_programs.rs:967]: (I think that github is having issues with PR, my comments keep appearing and disappearing, but since this disappeared again)

I think this sounds like we should use a BTreeMap with keys (pubkey, slot)? So we can key ranges and still have a total length. It's a tiny map so perf wise it's almost certainly better than juggling a two level hash table + vec.

File: program-runtime/src/loaded_programs.rs (hunk 6 above)
Comment (pgarg66) [reply program-runtime/src/loaded_programs.rs:1242]: Good catch. Update the test. Seeing this now.

```
$ cargo test -p solana-program-runtime -- test_random_eviction --nocapture
    Finished test [unoptimized + debuginfo] target(s) in 0.66s
     Running unittests src/lib.rs (target/debug/deps/solana_program_runtime-52501d36691531a3)

running 1 test
usage counter1 = 12, usage counter2 = 0
usage counter1 = 0, usage counter2 = 0
usage counter1 = 0, usage counter2 = 12
test loaded_programs::tests::test_random_eviction ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 35 filtered out; finished in 0.00s
```

File: program-runtime/src/loaded_programs.rs (hunk 7 above)
Comment (pgarg66) [reply program-runtime/src/loaded_programs.rs]: That's much better. Thanks.

File: program-runtime/src/loaded_programs.rs (hunk 2 above)
Comment (pgarg66) [reply program-runtime/src/loaded_programs.rs:959]: We do a reverse lookup for entries of a program during `extract()`. This helps find the most latest version of the program applicable to the given slot. That was one of the main reason/advantage of doing two level table.

Maybe this is not a big perf hit if we were to use BTreeMap. But, I think we should analyze it more carefully.

Comment (alessandrod) [reply program-runtime/src/loaded_programs.rs:940]: BTreeMap::range supports reverse iteration, I don't think it's going to be any slower than what we're currently doing?

File: program-runtime/src/loaded_programs.rs [hunk 8]
```diff
@@ -895,24 +934,62 @@ impl<FG: ForkGraph> LoadedPrograms<FG> {
...
+                .decayed_usage_counter(now);
+            (index, usage_counter)
+        }
+
+        for _ in 0..num_to_unload {
+            let (index1, usage_counter1) = random_index_and_usage_counter(&candidates, now);
+            let (index2, usage_counter2) = random_index_and_usage_counter(&candidates, now);
+
+            let (program, entry) = if usage_counter1 < usage_counter2 {
+                candidates.remove(index1)
```
Comment (alessandrod) [inline program-runtime/src/loaded_programs.rs:963]: We don't care about the order of candidates here right? If so we could make
these swap_remove and get O(1) instead of O(n) because of the memmoves

File: runtime/src/bank.rs [hunk 9]
```diff
@@ -5259,7 +5260,10 @@ impl Bank {
         self.loaded_programs_cache
             .write()
             .unwrap()
-            .sort_and_unload(Percentage::from(SHRINK_LOADED_PROGRAMS_TO_PERCENTAGE));
```
Comment (alessandrod) [inline runtime/src/bank.rs:5262]: sort_and_unload is now dead code and can probably be removed right?

Comment (pgarg66) [reply runtime/src/bank.rs:5262]: Left it in as an escape hatch. We can remove it once the new eviction policy is live and working on MB.

File: program-runtime/src/loaded_programs.rs (hunk 2 above)
Comment (pgarg66) [reply program-runtime/src/loaded_programs.rs:940]: We can/should certainly explore it. I believe it'll intersect with the other PR that does cooperative loading of programs.

Filed this issue to track it: https://github.com/solana-labs/solana/issues/34442

**Review by alessandrod — DISMISSED** [review]
lgtm

The flattening isn't great but I agree that it would make this diff considerably larger and it can be tried in a separate PR

**Comment by pgarg66** [comment]
> lgtm
> 
> The flattening isn't great but I agree that it would make this diff considerably larger and it can be tried in a separate PR

Thanks @alessandrod 
Leaving it open till tomorrow to give others a chance to take another look.

**Review by t-nelson — DISMISSED** [review]
we intend to let this ride 1.18 stabilization, right? no 1.17 bp

**Comment by pgarg66** [comment]
> we intend to let this ride 1.18 stabilization, right? no 1.17 bp

Yes, we are not backporting it to any of the branches. I am waiting to merge it to let other PRs go in, as it might create merge conflicts (or make it harder to backport the other PRs).

File: program-runtime/src/loaded_programs.rs [hunk 10]
```diff
@@ -935,6 +956,26 @@ impl<FG: ForkGraph> LoadedPrograms<FG> {
         })
     }
 
+    /// Returns the list of loaded programs which are verified and compiled.
+    fn get_flattened_entries(&self) -> Vec<(Pubkey, Arc<LoadedProgram>)> {
```
Comment (Lichtso) [inline program-runtime/src/loaded_programs.rs:960]: This could take two filter parameters `include_program_runtime_v1` and `include_program_runtime_v2`. Then `get_entries_sorted_by_tx_usage()`, which is only used by the recompilation phase could be inlined there.

File: program-runtime/src/loaded_programs.rs [hunk 11]
```diff
@@ -416,6 +423,20 @@ impl LoadedProgram {
...
+
+    pub fn update_access_slot(&self, slot: Slot) {
+        let _ = self.latest_access_slot.fetch_update(
+            Ordering::Relaxed,
+            Ordering::Relaxed,
+            |last_access| (last_access < slot).then_some(slot),
+        );
+    }
+
+    fn decayed_usage_counter(&self, now: Slot) -> u64 {
```
Comment (Lichtso) [inline program-runtime/src/loaded_programs.rs:435]: This should be used for the sorting at the beginning of the recompilation phase as well.

File: program-runtime/src/loaded_programs.rs [hunk 12]
```diff
@@ -416,6 +423,20 @@ impl LoadedProgram {
             && slot >= self.deployment_slot
             && slot < self.effective_slot
     }
+
+    pub fn update_access_slot(&self, slot: Slot) {
+        let _ = self.latest_access_slot.fetch_update(
+            Ordering::Relaxed,
+            Ordering::Relaxed,
+            |last_access| (last_access < slot).then_some(slot),
```
Comment (Lichtso) [inline program-runtime/src/loaded_programs.rs:431]: Looks like a manual implementation of [fetch_max](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicU64.html#method.fetch_max). But as I said above, I don't think atomics are necessary at all.

File: program-runtime/src/loaded_programs.rs [hunk 13]
```diff
@@ -862,7 +883,7 @@ impl<FG: ForkGraph> LoadedPrograms<FG> {
                             if let LoadedProgramType::Unloaded(_environment) = &entry.program {
                                 break;
                             }
-
+                            entry.update_access_slot(loaded_programs_for_tx_batch.slot);
```
Comment (Lichtso) [inline program-runtime/src/loaded_programs.rs:886]: Why only update it in this case here and not below where we also increment the usage counter?

File: program-runtime/src/loaded_programs.rs [hunk 14]
```diff
@@ -129,6 +130,8 @@ pub struct LoadedProgram {
     pub tx_usage_counter: AtomicU64,
     /// How often this entry was used by an instruction
     pub ix_usage_counter: AtomicU64,
+    /// Latest slot in which the entry was used
+    pub latest_access_slot: AtomicU64,
```
[the remark this answers is not in the document]
Comment (pgarg66) [reply program-runtime/src/loaded_programs.rs:134]: Will update. Before rebase, the extract did not take a `mut` reference to `self`.

File: program-runtime/src/loaded_programs.rs (hunk 13 above)
Comment (pgarg66) [reply program-runtime/src/loaded_programs.rs:886]: The test uses it too `test_usage_counter_decay`.

File: program-runtime/src/loaded_programs.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (pgarg66) [reply program-runtime/src/loaded_programs.rs:134]: Given that the entry is a cloned `Arc`, it'll require some sort of lock. So as discussed on Slack, sticking with Atomics for this.

[the remark this answers is not in the document]
Comment (Lichtso) [reply program-runtime/src/loaded_programs.rs:134]: I take it back, let's stick with atomics as we can not guarantee that we are the only owner. Other TX batches can still hold references, even if they never access these fields outside the global lock.

File: program-runtime/src/loaded_programs.rs (hunk 13 above)
Comment (pgarg66) [reply program-runtime/src/loaded_programs.rs:886]: Since we are sticking with Atomics, it'll be good to have a wrapper for test and the code.

File: program-runtime/src/loaded_programs.rs [hunk 15]
```diff
@@ -1004,13 +977,14 @@ impl<FG: ForkGraph> LoadedPrograms<FG> {
                         _ => None,
                     })
             })
-            .sorted_by_cached_key(|(_id, program)| program.tx_usage_counter.load(Ordering::Relaxed))
             .collect()
     }
 
     /// Unloads programs which were used infrequently
     pub fn sort_and_unload(&mut self, shrink_to: PercentageInteger) {
```
Comment (Lichtso) [inline program-runtime/src/loaded_programs.rs:984]: Is `sort_and_unload` still used somewhere?

Comment (pgarg66) [reply program-runtime/src/loaded_programs.rs:984]: No, left it in the code for time being, in case we need to go back to it. Let's get the new code to testnet and if it works well we can remove it.

**Review by Lichtso — APPROVED** [review]

File: program-runtime/src/loaded_programs.rs [hunk 16]
```diff
@@ -416,6 +422,16 @@ impl LoadedProgram {
...
             && slot < self.effective_slot
     }
+
+    pub fn update_access_slot(&self, slot: Slot) {
+        let _ = self.latest_access_slot.fetch_max(slot, Ordering::Relaxed);
+    }
+
+    pub fn decayed_usage_counter(&self, now: Slot) -> u64 {
+        let last_access = self.latest_access_slot.load(Ordering::Relaxed);
+        let decaying_for = now.saturating_sub(last_access);
```
Comment (Lichtso) [inline program-runtime/src/loaded_programs.rs:432]: `decaying_for` must be limited by `.min(64)` otherwise the line below might overflow on the bitshift.
